### ISSUE TITLE
Rewrite narrow stripes after RAIDZ expansion

### ISSUE THREAD

**Issue by TurkeyMan** [openzfs/zfs#15679]
### Describe the feature would like to see added to OpenZFS

I'd really love a tool to locate any narrow stripes and rewrite them at the new stripe width after expansion has completed.

### How will this feature improve OpenZFS?

RAIDZ expansion works by reordering the data on the drives; where existing data retains its original stripe width.
After expansion is complete, it may be desirable to reclaim disk space by rewriting old stripes at the new full stripe width.
There is no user-driven process to identify data written with the old stripe width and atomically rewrite the data with the new strip width. Work-arounds require copying data around the filesystem, they're not atomic, they also affect the file metadata/timestamps, etc.

**Comment by NicholasRush** [comment]
If ZFS could do this, it would mean that it has to rewrite all data on the expanded vdev. A better way would be to implement a raid-dp like raid model. In the raid-dp model, you can add everytime drives to an existing raid group without the need to rewrite everything. This is possible, because you have separate parity drives like raid-4.

**Comment by TurkeyMan** [comment]
> If ZFS could do this, it would mean that it has to rewrite all data on the expanded vdev.

Yes, it would. That's exactly the goal. A better way might be to do it _while_ expanding, but my understanding is that's not on the table; the expansion strategy was chosen because it's extremely safe and simple... (?)

**Comment by ericloewe** [comment]
This is not going to happen anytime soon since it would require block pointer rewrite.

**Comment by TheManchineel** [comment]
SnapRAID/Unraid it is 🤷

**Comment by TurkeyMan** [comment]
Great, how does it work? Do we run this as a second pass after expand completes?

I kinda imagined it might be integrated into a scrub at some level; these operations are extremely expensive, and finding ways to merge functionality and reduce the total number of passes is a massive win.

**Comment by amotin** [comment]
Scrub is not designed to modify anything. It only verifies checksums and recovers corrupted blocks.
